packages/toolpad-components/src/DataGrid.tsx

```diff
@@ -122,6 +131,9 @@ const DataGridComponent = React.forwardRef(function DataGridComponent(
 
   const columnsInitRef = React.useRef(false);
   const hasColumnsDefined = columnsProp && columnsProp.length > 0;
+
+  const idFieldInitRef = React.useRef(false);
```

Comment:
I don't think there's a need for refs, effects, debouncing... 

If my understanding is correct, then all this needs to do is derive some state for the rows from the props. Basically it's the following line that needs to be updated:

```tsx
const rows: GridRowsProp = rowsProp || dataQueryRows || EMPTY_ROWS;
```

It should become something along the lines of:

```tsx
  const rows: GridRowsProp = React.useMemo(() => {
    const parsedRows = rowsProp || dataQueryRows || EMPTY_ROWS;
    if (parsedRows.length <= 0 || rowIdField || parsedRows[0].id) {
      return parsedRows;
    }
    return parsedRows.map((row, id) => ({ ...rows, id }));
  }, [dataQueryRows, rowsProp, rowIdField]);
```